linkerd/proxy/transport/src/tls/client.rs

```diff
@@ -1,141 +1,84 @@
     L: HasConfig + Clone,
-    C: tower::Service<T, Response = TcpStream>,
+    C: tower::Service<T, Error = io::Error>,
+    C::Response: io::AsyncRead + io::AsyncWrite + io::PeerAddr + Send + Unpin + 'static,
     C::Future: Send + 'static,
-    C::Error: ::std::error::Error + Send + Sync + 'static,
-    C::Error: From<io::Error>,
 {
-    type Response = Connection;
-    type Error = C::Error;
-    type Future = ConnectFuture<L, C::Future>;
+    type Response = BoxedIo;
+    type Error = io::Error;
+    type Future = Either<Connect<C::Future, C::Response>, Handshake>;
 
+    #[inline]
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         self.inner.poll_ready(cx)
     }
 
     fn call(&mut self, target: T) -> Self::Future {
-        let peer_identity = target.peer_identity();
-        debug!(peer.identity = ?peer_identity);
-        let tls = self
-            .local
-            .clone()
-            .and_then(|l| peer_identity.map(|n| (n, l)));
-        ConnectFuture {
-            state: ConnectState::Init {
-                future: self.inner.call(target),
-                tls,
-            },
-        }
-    }
-}
-
-// ===== impl ConnectFuture =====
-
-impl<L, F> Future for ConnectFuture<L, F>
-where
-    L: HasConfig,
-    F: TryFuture<Ok = TcpStream>,
-    F::Error: From<io::Error>,
-{
-    type Output = Result<Connection, F::Error>;
-
-    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-        let mut this = self.project();
-        loop {
-            match this.state.as_mut().project() {
-                ConnectStateProj::Init { future, tls } => {
-                    let io = futures::ready!(future.try_poll(cx))?;
-
-                    match tls {
-                        Conditional::Some((peer_identity, local_tls)) => {
-                            trace!(peer.id = %peer_identity, "initiating TLS");
-                            let handshake =
-                                tokio_rustls::TlsConnector::from(local_tls.tls_client_config())
-                                    .connect(peer_identity.as_dns_name_ref(), io);
-                            this.state.set(ConnectState::Handshake(handshake));
-                        }
-                        Conditional::None(reason) => {
-                            trace!(%reason, "skipping TLS");
-                            return Poll::Ready(Ok(Connection::new(io)));
-                        }
-                    }
-                }
-                ConnectStateProj::Handshake(fut) => {
-                    let io = futures::ready!(fut.poll(cx))?;
-                    trace!("established TLS");
-                    return Poll::Ready(Ok(Connection::new(io)));
-                }
-            };
-        }
+        let tls = match self.local.clone() {
+            Conditional::Some(l) => tokio_rustls::TlsConnector::from(l.tls_client_config()),
+            Conditional::None(reason) => {
```

Comment:
super minor nit: we don't need to clone this in the case where identity is disabled, this could be
```suggestion
        let tls = match self.local {
            Conditional::Some(ref l) => tokio_rustls::TlsConnector::from(l.clone().tls_client_config()),
            Conditional::None(ref reason) => {
```
but, this probably doesn't make a meaningful difference, so who cares...

Reply:
Yeah, but in that case cloning the reason is cheap/trivial (it's just a marker type with no data).